### PR TITLE
generate server-side exceptions as classes

### PR DESCRIPTION
**Pull request by AllanZhengYP**
*Issue #, if available:*
Related to https://github.com/aws/aws-sdk-js-v3/pull/3261
SSDK generated example: https://github.com/aws/aws-sdk-js-v3/pull/3310

*Description of changes:*
This change generates server-side exceptions as individual classes. So customers can handle exceptions with ease. 
You can view generated code diff at https://github.com/aws/aws-sdk-js-v3/pull/3267

Now the customer can handle the errors without hardcoding the error code:
```javascript
import {
    InvalidSignatureException,
    UnfoundResourceException,
    FooServiceException
} from "@aws-sdk/client-foo";

try {
    await client.send(operation);
} catch (e) {
    if (e instanceof InvalidSignatureException) {
    // Do something
    } else if (e instanceof UnfoundResourceException) {
    // Do other thing
    } else if (e instanceof FooServiceException) {
    // Unknown exception
    } else {
    // SDK error
    }
}
```

For SSDK, customer can do the similar:
```javascript
import { InvalidSignatureException, UnfoundResourceException } from "@aws-sdk/server-foo";
try {
    await operation.handle(...);
} catch (e) {
    if (e instanceof InvalidSignatureException) {
    // Do something
    } else if (e instanceof UnfoundResourceException) {
    // Do other thing
    } else {
    // SDK error
    }
}
```

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by gosar** [comment]
The [commit for the generated code linked](https://github.com/aws/aws-sdk-js-v3/commit/2aa69b8ba047cbb03162ef7d7623cc0a3328670c) in the PR description says `SdkException`. Probably outdated link. It would be with `ServiceException`, right?

**Comment by AllanZhengYP** [comment]
@gosar Yes, they will be renamed to ServiceExceptions

File: smithy-typescript-codegen/src/main/java/software/amazon/smithy/typescript/codegen/StructureGenerator.java [hunk 1]
```diff
@@ -228,4 +172,65 @@ private void renderStructureNamespace(StructuredMemberWriter structuredMemberWri
...
+     * }
+     * }</pre>
+     *
+     * <p>The following TypeScript is generated:
+     *
+     * <pre>{@code
+     * import { ServiceException as __BaseException } from "@aws-sdk/smithy-client";
+     * import { ExceptionOptionType as __ExceptionOptionType } from "@aws-sdk/smithy-client";
+     *
+     * export interface NoSuchResource extends __BaseException {
```
Comment (adamthom-amzn) [inline smithy-typescript-codegen/src/main/java/software/amazon/smithy/typescript/codegen/StructureGenerator.java:198]: this should say class and not interface

**Review by adamthom-amzn — APPROVED** [review]

File: smithy-typescript-codegen/src/main/java/software/amazon/smithy/typescript/codegen/integration/AddBaseExceptionClassExport.java [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+
+import software.amazon.smithy.codegen.core.SymbolProvider;
+import software.amazon.smithy.model.Model;
+import software.amazon.smithy.typescript.codegen.TypeScriptDependency;
+import software.amazon.smithy.typescript.codegen.TypeScriptSettings;
+import software.amazon.smithy.typescript.codegen.TypeScriptWriter;
+import software.amazon.smithy.utils.SmithyInternalApi;
+
+/**
+ * Re-export the ServiceException class from index.ts.
```
Comment (gosar) [inline smithy-typescript-codegen/src/main/java/software/amazon/smithy/typescript/codegen/integration/AddBaseExceptionClassExport.java:26]: Need to update it for the base class used in ssdk too.

**Review by gosar — COMMENTED** [review]
Please update the link in the description of the PR with the latest generated code

File: smithy-typescript-codegen/src/main/java/software/amazon/smithy/typescript/codegen/IndexGenerator.java [hunk 3]
```diff
@@ -82,6 +82,9 @@ static void writeServerIndex(
         writer.write("export * from \"./operations\";");
 
         writer.write("export * from \"./$L\"", symbol.getName());
+
+        writer.write("export { SmithyException } from $S", TypeScriptDependency.SERVER_COMMON.packageName);
```
Comment (gosar) [inline smithy-typescript-codegen/src/main/java/software/amazon/smithy/typescript/codegen/IndexGenerator.java:86]: Maybe there's a reason I'm not aware of, but why is this here v/s AddBaseExceptionClassExport for the re-export on client side? If they can be same, seems like having it here is better v/s a new TypeScriptIntegration.

File: smithy-typescript-codegen/src/main/java/software/amazon/smithy/typescript/codegen/StructureGenerator.java [hunk 4]
```diff
@@ -228,4 +172,65 @@ private void renderStructureNamespace(StructuredMemberWriter structuredMemberWri
             });
         });
     }
+
+    /**
+     * Error structures generate interfaces that extend from SmithyException
```
Comment (gosar) [inline smithy-typescript-codegen/src/main/java/software/amazon/smithy/typescript/codegen/StructureGenerator.java:177]: generate classes that extend from ServiceException

File: smithy-typescript-codegen/src/main/java/software/amazon/smithy/typescript/codegen/StructureGenerator.java [hunk 5]
```diff
@@ -228,4 +172,65 @@ private void renderStructureNamespace(StructuredMemberWriter structuredMemberWri
...
+        writer.openBlock("export class $L extends $L {", symbol.getName(), "__BaseException");
+        writer.write("readonly name: $1S = $1S;", shape.getId().getName());
+        writer.write("readonly $$fault: $1S = $1S;", errorTrait.getValue());
+        if (!isServerSdk) {
+            HttpProtocolGeneratorUtils.writeRetryableTrait(writer, shape, ";");
+        }
+        StructuredMemberWriter structuredMemberWriter = new StructuredMemberWriter(model, symbolProvider,
+                shape.getAllMembers().values());
+        // since any error interface must extend from JavaScript Error interface, message member is already
+        // required in the JavaScript Error interface
```
Comment (gosar) [inline smithy-typescript-codegen/src/main/java/software/amazon/smithy/typescript/codegen/StructureGenerator.java:229]: nit: should it say Error class?

File: smithy-typescript-ssdk-libs/server-common/src/errors.ts [hunk 6]
```diff
@@ -29,37 +41,57 @@ export const isFrameworkException = (error: any): error is SmithyFrameworkExcept
   return error.$frameworkError;
 };
 
-export class InternalFailureException implements SmithyException {
+export class InternalFailureException extends SmithyException {
   readonly name = "InternalFailure";
   readonly $fault = "server";
   readonly statusCode = 500;
```
Comment (gosar) [inline smithy-typescript-ssdk-libs/server-common/src/errors.ts:47]: I think we decided we don't need this. Though could be done is separate PR if that's what you had planned.

File: smithy-typescript-codegen/src/main/java/software/amazon/smithy/typescript/codegen/IndexGenerator.java (hunk 3 above)
Comment (AllanZhengYP) [reply smithy-typescript-codegen/src/main/java/software/amazon/smithy/typescript/codegen/IndexGenerator.java:86]: The `writeServerIndex()` doesn't call the customer TypeScriptIntegration. Generally the TypeScriptIntegration interface is not fully supported in SSDK. So I leave them here. But I can apply the `writeAdditionalExports()` of `TypeScriptIntegration` for SSDK if you want.

File: smithy-typescript-ssdk-libs/server-common/src/errors.ts [hunk 7]
```diff
@@ -13,7 +13,19 @@
  *  permissions and limitations under the License.
  */
 
-import { SmithyException } from "@aws-sdk/smithy-client";
+export class SmithyException extends Error {
```
Comment (gosar) [inline smithy-typescript-ssdk-libs/server-common/src/errors.ts:16]: Can we call this ServiceException now?

**Review by trivikr — APPROVED** [review]
